estimatefee.go

```diff
@@ -0,0 +1,434 @@
+
+	// The cached estimates.
+	cached []float64
+
+	// Transactions that have been removed from the bins. This allows us to
+	// revert in case of an orphaned block.
+	dropped []registeredBlock
+}
+
+// newFeeEstimator creates a feeEstimator for which at most maxRollback blocks
+// can be unregistered and which returns an error unless minRegisteredBlocks
+// have been registered with it.
+func newFeeEstimator(maxRollback, minRegisteredBlocks uint32) *feeEstimator {
+	return &feeEstimator{
+		maxRollback:         maxRollback,
+		minRegisteredBlocks: minRegisteredBlocks,
+		lastKnownHeight:     mempoolHeight,
+		binSize:             estimateFeeBinSize,
+		maxReplacements:     estimateFeeMaxReplacements,
+		observed:            make(map[chainhash.Hash]observedTransaction),
+		dropped:             make([]registeredBlock, 0, maxRollback),
+	}
+}
+
+// ObserveTransaction is called when a new transaction is observed in the mempool.
+func (ef *feeEstimator) ObserveTransaction(t *mempoolTxDesc) {
+	ef.Lock()
+	defer ef.Unlock()
+
+	hash := *t.Tx.Hash()
+	if _, ok := ef.observed[hash]; !ok {
+		size := uint32(t.Tx.MsgTx().SerializeSize())
+		fee := uint64(t.Fee)
+
+		ef.observed[hash] = observedTransaction{
+			hash:     hash,
+			size:     size,
+			fee:      fee,
+			feePerKb: float64(1000*fee) / float64(size),
+			observed: t.Height,
+			mined:    mempoolHeight,
+		}
+	}
+}
+
+// RegisterBlock informs the fee estimator of a new block to take into account.
+func (ef *feeEstimator) RecordBlock(block *btcutil.Block) {
+	ef.Lock()
+	defer ef.Unlock()
+
+	// The previous sorted list is invalid, so delete it.
+	ef.cached = nil
+
+	height := block.Height()
+	if height != ef.lastKnownHeight+1 && ef.lastKnownHeight != mempoolHeight {
+		panic(fmt.Sprint("intermediate block not recorded; current height is ", ef.lastKnownHeight,
+			"; new height is ", height))
+	}
+
+	// Update the last known height.
+	ef.lastKnownHeight = height
+	ef.numBlocksRegistered++
+
+	// Randomly order txs in block.
+	transactions := make(map[*btcutil.Tx]struct{})
+	for _, t := range block.Transactions() {
+		transactions[t] = struct{}{}
+	}
+
+	// Count the number of replacements we make per bin so that we don't
+	// replace too many.
+	var replacementCounts [estimateFeeBins]int
+
+	// Keep track of which txs were dropped in case of an orphan block.
+	dropped := registeredBlock{
+		hash:         block.Hash(),
+		transactions: make([]*observedTransaction, 0, 100),
+	}
+
+	// Go through the txs in the block.
+	for t := range transactions {
+		hash := *t.Hash()
+
+		// Have we observed this tx in the mempool?
+		o, ok := ef.observed[hash]
+		if !ok {
+			continue
+		}
+
+		// Put the observed tx in the oppropriate bin.
+		o.mined = height
+
+		blocksToConfirm := height - o.observed - 1
+
+		// Make sure we do not replace too many transactions per min.
+		if replacementCounts[blocksToConfirm] == ef.maxReplacements {
```

Comment:
That shouldn't happen because all transactions that have been observed more than 25 blocks ago are removed after a block is registered. Once a new block comes in, there shouldn't be any which observedTransactions which have taken > 25 blocks to confirm. I'll add a check in there anyway.